Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1186,123 +1186,127 @@ $ adb shell /data/local/tmp/validate 12345
+│     ╭───< 0x00001814      bne 0x1854                                 ; likely
+│     │││   0x00001818      ldrb r4, [fp, -0x19]
+│     │││   0x0000181c      bl fcn.00001728
+│     │││   0x00001820      mov r3, r0
+│     │││   0x00001824      cmp r4, r3
+│    ╭────< 0x00001828      bne 0x1854                                 ; likely
+│    ││││   0x0000182c      ldrb r4, [fp, -0x18]
+│    ││││   0x00001830      bl fcn.00001744
+│    ││││   0x00001834      mov r3, r0
+│    ││││   0x00001838      cmp r4, r3
+│   ╭─────< 0x0000183c      bne 0x1854                                 ; likely
+│   │││││   0x00001840      ldr r3, [0x0000186c]                       ; [0x186c:4]=0x270 section..hash ; section..hash
+│   │││││   0x00001844      add r3, pc, r3                             ; 0x1abc ; "Product activation passed. Congratulations!"
+│   │││││   0x00001848      mov r0, r3                                 ; 0x1abc ; "Product activation passed. Congratulations!" ;
+│   │││││   0x0000184c      bl sym.imp.puts                            ; int puts(const char *s)
+│   │││││                                                              ; int puts("Product activation passed. Congratulations!")
+│  ╭──────< 0x00001850      b 0x1864
+│  ││││││   ; CODE XREFS from fcn.00001760 (0x17ec, 0x1800, 0x1814, 0x1828, 0x183c)
+│  │╰╰╰╰╰─> 0x00001854      ldr r3, aav.0x00000288                     ; [0x1870:4]=0x288 aav.0x00000288
+│  │        0x00001858      add r3, pc, r3                             ; 0x1ae8 ; "Incorrect serial." ;
+│  │        0x0000185c      mov r0, r3                                 ; 0x1ae8 ; "Incorrect serial." ;
+│  │        0x00001860      bl sym.imp.puts                            ; int puts(const char *s)
+│  │                                                                   ; int puts("Incorrect serial.")
+│  │        ; CODE XREF from fcn.00001760 (0x1850)
+│  ╰──────> 0x00001864      sub sp, fp, 8
+╰           0x00001868      pop {r4, fp, pc}                           ; entry.preinit0 ; entry.preinit0 ;
+```
+
+If you look in the graph view you can see a loop with some XOR-magic happening at 0x00001784, which supposedly decodes the input string.
+
+![Loop](Images/Chapters/0x05c/loop_1784.png)
+
+Starting from 0x000017dc, you can see a series of decoded values compared with values from further subfunction calls.
+
+![Decoded values being compared](Images/Chapters/0x05c/values_compare_17dc.png)
+
+Even though this doesn't look like highly sophisticated stuff, you'd still need to analyze more to completely reverse this check and generate a license key that passes it. Now comes the twist: dynamic symbolic execution enables you to construct a valid key automatically! The symbolic execution engine maps a path between the first instruction of the license check (0x00001760) and the code that prints the "Product activation passed" message (0x00001840) to determine the constraints on each byte of the input string.
```

Comment:
Even though this doesn't look like highly sophisticated stuff,  -> Even though this doesn't look highly sophisticated,

Reply:
Done, thanks!